docs/design/readme.md

```diff
@@ -0,0 +1,1704 @@
+- Provides a convenient interface for invoking and serving RPC service calls.
+
+**`libcyphal::application`** is a layer on top of the previous two that provides the top-level node-centric API for the application and implements some common functions of the protocol, such as the register interface, publication of heartbeats and port usage messages, and so on. This is where the included batteries can be found. Unlike the lower layers, the application layer depends on the code generated by Nunavut from certain standard DSDL definitions.
+
+## General considerations
+
+### Design principles
+
+Exceptions are not leveraged by the library itself, but it is designed to be exception-aware, allowing its integration into applications that make use of exception handling.
+
+Runtime type identification is not required but allowed as an optional feature to facilitate dynamic linking when used in a high-level operating system; more on this later.
+
+Usage of the heap is reduced to the bare minimum and is done strictly via polymorphic memory resources without any reliance on the global heap. Most allocations are done once during initialization; with the exceptions listed below, the library will not perform further allocations at runtime unless the application needs to change its port configuration (i.e., subscriptions/publications/RPC) dynamically at runtime, which is rarely seen in embedded real-time systems. The exceptions to the no-memory-allocation policy are as follows:
+
+- Transport-layer queues and memory reassembly buffers allocated at runtime by the lizards. The memory usage patterns and limits of the lizards are very well formalized in the documentation; further, certain lizards allow the replacement of heap allocators with block pool allocators.
+
+Type erasure is used throughout to reduce coupling and uphold encapsulation. In some cases, heap allocation that is often associated with type erasure is avoided with the help of small-object-optimized containers that provide an internal arena to store the type-erased object within. More on this below.
+
+Unconventionally, dynamic typing is leveraged with care in certain platform-specific contexts to uphold encapsulation.
+
+### RTTI and dynamic linking considerations
+
+As will be seen later, certain design features are dependent on the ability to detect identical types at runtime, yet the library must be usable without the RTTI capability of the C++ runtime. The following construct is used to provide a small subset of RTTI features that happens to be sufficient for the needs of the library:
+
+```c++
+/// A simplified substitute for RTTI that allows querying the identity of types.
+/// Returns an object of an unspecified type for which operator== and operator!= are defined.
+/// Objects for the same type compare equal. Ordering is not defined.
+template <typename>
+auto getTypeID() noexcept
+{
+    static const struct {} placeholder{};
+    return static_cast<const void*>(&placeholder);
```

Comment:
Should return a `void const * const` since it's static. or just `void * const` if you are not trying to return a constant void but are trying to return a constant pointer.

Reply:
cv-qualifiers don't survive copying, so returning `const T` is equivalent to returning `T`. Here, `T = const void*`.